File: backend/internal/domain/room_geometry.go
package domain

// RoomGeometry (Layer 1) represents the physical environment.

type RoomGeometry struct {
	ID        string     `json:"id"`
	Name      string     `json:"name,omitempty"`
	Perimeter Polygon2D  `json:"perimeter"`
	Floor     Plane      `json:"floor"`
	Ceiling   Plane      `json:"ceiling"`
	Walls     []Wall     `json:"walls"`
	Openings  []Opening  `json:"openings"`
	Obstacles []Obstacle `json:"obstacles"`
}

type Wall struct {
	ID          string    `json:"id"`
	Vertices    []Point3D `json:"vertices"`
	Thickness   float64   `json:"thickness"`
	OutOfPlumb  *Point3D  `json:"outOfPlumb,omitempty"`
}

type OpeningType string

const (
	OpeningDoor   OpeningType = "door"
	OpeningWindow OpeningType = "window"
)

type Opening struct {
	ID       string      `json:"id"`
	Type     OpeningType `json:"type"`
	WallID   string      `json:"wallId"`
	Origin   Point3D     `json:"origin"`
	Width    float64     `json:"width"`
	Height   float64     `json:"height"`
}

type ObstacleType string

const (
	ObstacleColumn   ObstacleType = "column"
	ObstacleBeam     ObstacleType = "beam"
	ObstaclePipe     ObstacleType = "pipe"
	ObstacleSkirting ObstacleType = "skirting"
	ObstacleOther    ObstacleType = "other"
)

type Obstacle struct {
	ID       string       `json:"id"`
	Type     ObstacleType `json:"type"`
	Label    string       `json:"label,omitempty"`
	Bounds   BoundingBox3D `json:"bounds"`
	Profile  []Point3D    `json:"profile,omitempty"`
}

File: frontend/src/domain/room-geometry.ts
import type { Plane, Point3D, Polygon2D } from './geometry';

/**
 * Layer 1 — Room Geometry
 * Represents the physical environment. Independent from furniture definition.
 */

export interface RoomGeometry {
  id: string;
  name?: string;
  perimeter: Polygon2D;
  floor: Plane;
  ceiling: Plane;
  walls: Wall[];
  openings: Opening[];
  obstacles: Obstacle[];
}

export interface Wall {
  id: string;
  /** Vertices along the wall face (typically bottom then top edge). */
  vertices: Point3D[];
  thickness: number;
  /** Deviation from vertical expressed as a vector, not an angle. */
  outOfPlumb?: Point3D;
}

export type OpeningType = 'door' | 'window';

export interface Opening {
  id: string;
  type: OpeningType;
  wallId: string;
  /** Bottom-left corner of the opening on the wall face. */
  origin: Point3D;
  width: number;
  height: number;
}

export type ObstacleType =
  | 'column'
  | 'beam'
  | 'pipe'
  | 'skirting'
  | 'other';

export interface Obstacle {
  id: string;
  type: ObstacleType;
  label?: string;
  /** Obstacle bounds in room coordinates. */
  bounds: { min: Point3D; max: Point3D };
  /** Optional profile vertices for non-rectangular obstacles. */
  profile?: Point3D[];
}
